ambry-router/src/main/java/com/github/ambry/router/QuotaAwareOperationController.java

```diff
@@ -0,0 +1,188 @@
+/**
+ * Copyright 2021 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.router;
+
+import com.github.ambry.account.AccountService;
+import com.github.ambry.clustermap.ClusterMap;
+import com.github.ambry.commons.ResponseHandler;
+import com.github.ambry.config.RouterConfig;
+import com.github.ambry.network.NetworkClient;
+import com.github.ambry.network.NetworkClientFactory;
+import com.github.ambry.network.RequestInfo;
+import com.github.ambry.notification.NotificationSystem;
+import com.github.ambry.quota.QuotaResource;
+import com.github.ambry.quota.QuotaResourceType;
+import com.github.ambry.utils.Time;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * {@link OperationController} implementation that dispatches chunk requests to server based on quota.
+ */
+public class QuotaAwareOperationController extends OperationController {
+  private static final Logger logger = LoggerFactory.getLogger(QuotaAwareOperationController.class);
+  // QuotaResource instance to use for cases where OperationController gets null QuotaResource.
+  private static final QuotaResource UNKNOWN_QUOTA_RESOURCE = new QuotaResource("UNKNOWN", QuotaResourceType.ACCOUNT);
+  private final Map<QuotaResource, LinkedList<RequestInfo>> requestQueue = new HashMap<>();
+  private final NonBlockingRouter nonBlockingRouter;
+  private final NonBlockingRouterMetrics routerMetrics;
```

Comment:
Seems like member variables `nonBlockingRouter` and `routerMetrics` are also defined in base class `OperationController`.  May be reuse them?

Reply:
done